flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/Bucket.java

```diff
@@ -53,175 +58,260 @@
 
 	private final RecoverableWriter fsWriter;
 
-	private final Map<Long, List<RecoverableWriter.CommitRecoverable>> pendingPerCheckpoint = new HashMap<>();
-
-	private long partCounter;
-
-	private PartFileWriter<IN, BucketID> currentPart;
+	private final RollingPolicy<IN, BucketID> rollingPolicy;
 
-	private List<RecoverableWriter.CommitRecoverable> pending;
-
-	/**
-	 * Constructor to restore a bucket from checkpointed state.
-	 */
-	public Bucket(
-			RecoverableWriter fsWriter,
-			int subtaskIndex,
-			long initialPartCounter,
-			PartFileWriter.PartFileFactory<IN, BucketID> partFileFactory,
-			BucketState<BucketID> bucketState) throws IOException {
+	private final Map<Long, List<RecoverableWriter.CommitRecoverable>> pendingPartsPerCheckpoint = new HashMap<>();
 
-		this(fsWriter, subtaskIndex, bucketState.getBucketId(), bucketState.getBucketPath(), initialPartCounter, partFileFactory);
-
-		// the constructor must have already initialized the filesystem writer
-		Preconditions.checkState(fsWriter != null);
-
-		// we try to resume the previous in-progress file, if the filesystem
-		// supports such operation. If not, we just commit the file and start fresh.
+	private long partCounter;
 
-		final RecoverableWriter.ResumeRecoverable resumable = bucketState.getInProgress();
-		if (resumable != null) {
-			currentPart = partFileFactory.resumeFrom(
-					bucketId, fsWriter, resumable, bucketState.getCreationTime());
-		}
+	private PartFileWriter<IN, BucketID> inProgressPart;
 
-		// we commit pending files for previous checkpoints to the last successful one
-		// (from which we are recovering from)
-		for (List<RecoverableWriter.CommitRecoverable> commitables: bucketState.getPendingPerCheckpoint().values()) {
-			for (RecoverableWriter.CommitRecoverable commitable: commitables) {
-				fsWriter.recoverForCommit(commitable).commitAfterRecovery();
-			}
-		}
-	}
+	private List<RecoverableWriter.CommitRecoverable> pendingPartsForCurrentCheckpoint;
 
 	/**
 	 * Constructor to create a new empty bucket.
 	 */
-	public Bucket(
-			RecoverableWriter fsWriter,
-			int subtaskIndex,
-			BucketID bucketId,
-			Path bucketPath,
-			long initialPartCounter,
-			PartFileWriter.PartFileFactory<IN, BucketID> partFileFactory) {
+	private Bucket(
+			final RecoverableWriter fsWriter,
+			final int subtaskIndex,
+			final BucketID bucketId,
+			final Path bucketPath,
+			final long initialPartCounter,
+			final PartFileWriter.PartFileFactory<IN, BucketID> partFileFactory,
+			final RollingPolicy<IN, BucketID> rollingPolicy) {
 
 		this.fsWriter = Preconditions.checkNotNull(fsWriter);
 		this.subtaskIndex = subtaskIndex;
 		this.bucketId = Preconditions.checkNotNull(bucketId);
 		this.bucketPath = Preconditions.checkNotNull(bucketPath);
 		this.partCounter = initialPartCounter;
+		this.rollingPolicy = Preconditions.checkNotNull(rollingPolicy);
```

Comment:
this is very much personal preference, but when I'm just assigning stuff I do that in the same order as the constructors arguments appear.